[CBRD-26274] 멀티 ALTER TABLE 을 이용한 PK 변경 허용

youngjun9072 opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-26274>](http://jira.cubrid.org/browse/CBRD-26274)

### Purpose

HA 환경에서는 do_alter_one_clause_with_template() 함수에서 멀티 ALTER TABLE 을 이용해 PK 를 변경할 수 없습니다.

이는 DROP PRIMARY KEY를 수행한 직후 HA에 대한 제약조건을 확인하는 코드가 수행되기때문입니다.

이를 해결하기위해 HA 제약사항 확인 함수를 do_alter_one_clause_with_template()함수의 상위 함수인

do_alter() 함수로 이동합니다.

### Implementation

do_alter() 함수에서 alter에 대한 모든 clause 를 확인 한 후 수행 완료된 vclass에 대한 HA 제약사항을 확인하도록 수정합니다.

### Remarks
1. 위치상 HA제약과 관련 없는 작업에도 매번 수행되어야하는 루틴이지만 alter table 은 빈번한 호출이 아니라 괜찮을거란 주간 회의중 의견이 있었습니다. 

2. check_ha_repl_constraint() 함수를 호출하기위해 vclass() 도 함께 호출되고있는데 
이 부분을 check_ha_repl_constraint_with_name(char * entity_name) 과같은 함수로 한번 더 감쌀지 고민중입니다. 의견있으시면 부탁드립니다.

youngjun9072 committed:
Refactor HA constraint verification to support multiple ALTER TABLE clauses

- Moved replication key (HA constraint) check from do_alter_one_clause_with_template() to do_alter()
  to properly handle multiple ALTER TABLE clauses within a single statement.
- Ensures HA constraint verification is performed once after all clauses are processed.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1733,6 +1727,15 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
+  if (error_code != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_code, 0);
+      return error_code;
```
```suggestion
      goto error_exit;
```

youngjun9072 committed:
Fix HA constraint check missing after ALTER TABLE ... RENAME

- Corrected a bug where HA replication constraint validation was skipped
  when processing ALTER TABLE statements containing RENAME clauses.
  The final table name is now tracked properly to ensure validation is applied.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1733,6 +1734,14 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
       do_semantic_checks = true;
     }
 
+  vclass = db_find_class (entity_name);
```
모든 alter_code에 대해 허용하기 보다는 범위를 제약하는게 어떨까요? bool need_check_repl_constraint flag 변수를 이용해서 이 변수가 설정된 경우에만 검사하도록 한다면, PT_RENAME_ENTITY, PT_CHANGE_AUTO_INCREMENT, PT_CHANGE_COLLATION 등의 변경에 대해서 제약 검사를 수행할 필요가 없고, 또한 로직 상 코드 이해도가 보장될 것 같습니다.

youngjun9072 replied:
@hornetmj 반영하였습니다. ha제약사항 확인 대상이 되는 연산 코드는
1. PK및 UK제거: PT_DROP_CONSTRAINT, PT_DROP_PRIMARY_CLAUSE 
2. 외래키 어트리뷰트 추가: PT_ADD_ATTR_MTHD
3. RK 컬럼 제거: PT_DROP_ATTR_MTHD
4. ALTER TABLE을 이용한 RENAME: PT_RENAME_ENTITY
입니다.

youngjun9072 committed:
Validate HA replication constraints only when required via need_check_repl_constraint

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1733,6 +1748,17 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
+      error_code = check_ha_repl_constraint (vclass);
+      if (error_code != NO_ERROR)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error_code, 0);
```
에러가 발생하는 check_ha_repl_constraint() 함수 내에서 er_set()을 설정하시는게 좋을 것 같습니다. 그러면, check_ha_repl_constraint() 호출자에서 매번 er_set()을 수행할 필요가 없어집니다.

youngjun9072 replied:
@hornetmj  추후 에러 메시지 변경 이슈나 작은 이슈들 처리할때 한번에 묶어 수정하는게 어떨까요? 이번 이슈와 상관없은 do_create_entity에도 변경사항이 발생됩니다.

hornetmj replied:
넵. TODO로 이력 남겨주세요.

youngjun9072 replied:
@hornetmj epic에 todo로 기록해뒀습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1733,6 +1748,17 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
       do_semantic_checks = true;
     }
 
+  if (error_code == NO_ERROR && need_check_repl_constraint)
```
```suggestion
  if (need_check_repl_constraint)
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1721,7 +1736,7 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
 	  crt_clause->next = NULL;
 
-	  error_code = do_alter_one_clause_with_template (parser, crt_clause);
+	  error_code = do_alter_one_clause_with_template (parser, crt_clause, &need_check_repl_constraint);
```
NIT:

do_alter_one_clause_with_template() 함수로 need_check_repl_constraint를 넘기기 보다는 다음과 같은 방식이 더 직관적이고, 추후 유지보수가 용이할 것으로 판단되어 의견드립니다.

```
for (crt_clause = alter; crt_clause != NULL; crt_clause = crt_clause->next)
{
    ...
    switch (alter_code)
    {
        ...
    }

    // 이 부분은 매크로 or 함수로 작성 검토
    if (need_check_repl_constraint == false)
    {
        switch (alter_code)
            case PT_ADD_ATTR_MTHD:
            case PT_DROP_ATTR_MTHD:
            case PT_DROP_CONSTRAINT:
            case PT_DROP_PRIMARY_CLAUSE:
            ...
                need_check_repl_constraint = true;
    }

    ...
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -1687,7 +1694,15 @@ do_alter (PARSER_CONTEXT * parser, PT_NODE * alter)
+	   */
+	  entity_name = crt_clause->info.alter.alter_clause.rename.new_name->info.name.original;
 	  error_code = do_alter_clause_rename_entity (parser, crt_clause);
+	  need_check_repl_constraint = true;
```
rename에 의해서 변경된 entity_name을 할당(기억)하는 부분은 필요한 작업이라고 생각됩니다. 궁금한 점은 need_check_repl_constraint = true 설정은 필요하지 않을 것 같은데, 혹시 필요한 이유가 있을까요?

youngjun9072 replied:
@hornetmj 잘못 작성된 코드입니다. 제거했습니다.

youngjun9072 committed:
Apply review feedback

- Moved the routine for flagging `need_check_repl_constraint` to do_alter()
  and introduced the `IS_REPL_CONSTRAINT_RELATED_ALTER` macro.
- Removed the unnecessary `need_check_repl_constraint` check in the
  PT_RENAME_ENTITY clause.
